leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/ClientRegistryImpl.java

```diff
@@ -67,22 +67,32 @@ public Client get(String endpoint) {
     }
 
     @Override
-    public boolean registerClient(Client client) {
-        Validate.notNull(client);
+    public boolean isRegisterAllowed(final Client client)
+    {
+       Validate.notNull(client);
+       LOG.debug("checking if Client {} is allowed to register", client);
+       clientsByEp.put(client.getEndpoint(), client);
```

Comment:
You got it right, the nasty side effect here is, that the unregister() will always happen once before the register(), because 'previous' is set. And the listeners have to deal with that...